File: mesh/android/ExoPlayerMeshJava/app/src/main/java/io/streamroot/ctl/delivery/client/mesh/exoplayermesh/PlayerInteractor.java
package io.streamroot.ctl.delivery.client.mesh.exoplayermesh;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.Nullable;

import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.LoadControl;
import com.google.android.exoplayer2.PlaybackException;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;

import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import io.streamroot.lumen.delivery.client.core.LumenPlayerInteractorBase;
import io.streamroot.lumen.delivery.client.core.LumenVideoPlaybackState;

public final class ExoPlayerInteractor extends LumenPlayerInteractorBase implements Player.Listener {

    private interface BufferTargetBridge {
        double bufferTarget();
        void setBufferTarget(double bufferTarget);
    }

    private static abstract class LoadControlBufferTargetBridge implements BufferTargetBridge {
        private static final String MAX_BUFFER_FIELD_NAME = "maxBufferUs";

        private final LoadControl loadControl;

        private final Field maxBufferField;
        private final long minBufferUs;

        private LoadControlBufferTargetBridge(LoadControl loadControl, long minBufferUs) {
            this.loadControl = loadControl;
            maxBufferField = getAccessibleFieldElseThrow(loadControl.getClass(), MAX_BUFFER_FIELD_NAME);
            this.minBufferUs = minBufferUs;
        }

        private static Field getAccessibleFieldElseThrow(Class<?> clazz, String fieldName) {
            try {
                final Field minBufferField = clazz.getDeclaredField(fieldName);
                minBufferField.setAccessible(true);
                return minBufferField;
            } catch (Exception e) {
                throw new IllegalArgumentException(String.format(
                        "Impossible to retrieve field `%s` value from LoadControl of type `%s`",
                        fieldName, LoadControl.class.getSimpleName()));
            }
        }

        static long getLongFromFieldElseThrow(LoadControl loadControl, String fieldName) {
            try {
                return getAccessibleFieldElseThrow(loadControl.getClass(), fieldName).getLong(loadControl);
            } catch (Exception e) {
                throw new IllegalArgumentException(String.format(
                        "Impossible to retrieve long field `%s` value from LoadControl of type `%s`",
                        fieldName, LoadControl.class.getSimpleName()));
            }
        }

        @Override
        public double bufferTarget() {
            try {
                return (double) TimeUnit.MICROSECONDS.toSeconds(maxBufferField.getLong(loadControl));
            } catch (Exception e) {
                return 0.0;
            }
        }

        @Override
        public void setBufferTarget(double bufferTarget) {
            final long maxBufferUs = TimeUnit.SECONDS.toMicros((long)bufferTarget);
            if (maxBufferUs >= minBufferUs) try {
                maxBufferField.setLong(
                        loadControl,
                        maxBufferUs
                );
            } catch (Exception ignored) {}
        }
    }

    // 2.10- & 2.12+
    private static final class LoadControlBufferTargetBridgeV1 extends LoadControlBufferTargetBridge {
        private static final String MIN_BUFFER_FIELD_NAME = "minBufferUs";

        private LoadControlBufferTargetBridgeV1(LoadControl loadControl) {
            super(
                    loadControl,
                    getLongFromFieldElseThrow(loadControl, MIN_BUFFER_FIELD_NAME)
            );
        }
    }

    // 2.11
    private static final class LoadControlBufferTargetBridgeV2 extends LoadControlBufferTargetBridge {
        private static final String MIN_BUFFER_AUDIO_FIELD_NAME = "minBufferAudioUs";
        private static final String MIN_BUFFER_VIDEO_FIELD_NAME = "minBufferVideoUs";

        private LoadControlBufferTargetBridgeV2(LoadControl loadControl, boolean audioOnly) {
            super(
                    loadControl,
                    getLongFromFieldElseThrow(
                            loadControl,
                            audioOnly ? MIN_BUFFER_AUDIO_FIELD_NAME : MIN_BUFFER_VIDEO_FIELD_NAME
                    )
            );
        }
    }

    private static final class BufferTargetBridgeFactory {
        private static BufferTargetBridge createInteractor(LoadControl loadControl, boolean audioOnly) {
            try {
                final LoadControlBufferTargetBridge bridge = new LoadControlBufferTargetBridgeV1(loadControl);
                Log.v("Misc", "Using interactor V1");
                return bridge;
            } catch (Exception ignored) {
                try {
                    final LoadControlBufferTargetBridge bridge = new LoadControlBufferTargetBridgeV2(loadControl, audioOnly);
                    Log.v("Misc", "Using interactor V2");
                    return bridge;
                } catch (Exception e) {
                    throw new RuntimeException("Unsupported ExoPlayer version");
                }
            }
        }
    }

    private final Handler handler;

    private final ExoPlayer player;
    final LoadControl loadControl;
    private final ExoPlayerBandwidthMeter bandwidthMeter;

    // Conversion from milliseconds to seconds does not use TimeUnit because we want
    // to keep a floating precision
    final private BufferTargetBridge bridge;

    public ExoPlayerInteractor(ExoPlayer player, LoadControl loadControl, ExoPlayerBandwidthMeter bandwidthMeter) {
        this.handler = new Handler(player.getApplicationLooper());
        this.player = player;
        this.loadControl = loadControl;
        this.bandwidthMeter = bandwidthMeter;
        this.bridge = BufferTargetBridgeFactory.createInteractor(loadControl, false);

        player.addListener(this);
    }

    @Override
    public void onSeekProcessed() {
        super.playerStateChange(LumenVideoPlaybackState.SEEKING);
    }

    @Override
    public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {
        super.playerTrackSwitch();
    }

    @Override
    public void onPlayerError(PlaybackException error) {
        super.playerError();
    }

    @Override
    public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
        switch (playbackState) {
            case Player.STATE_IDLE:
                super.playerStateChange(LumenVideoPlaybackState.IDLE);
                break;
            case Player.STATE_BUFFERING:
                super.playerStateChange(LumenVideoPlaybackState.REBUFFERING);
                break;
            case Player.STATE_READY:
                super.playerStateChange(playWhenReady ? LumenVideoPlaybackState.PLAYING : LumenVideoPlaybackState.PAUSED);
                break;
            case Player.STATE_ENDED:
                super.playerStateChange(LumenVideoPlaybackState.ENDED);
                break;
        }
    }

    private interface Function<T> {
        T execute();
    }

    private <T> T runSyncOnEPHandler(final Function<T> block) throws InterruptedException {
        final Object lock = new Object();
        final AtomicReference<T> tRef = new AtomicReference<>(null);
        synchronized (lock) {
            handler.post(() -> {
                synchronized (lock) {
                    tRef.set(block.execute());
                    lock.notify();
                }
            });
            lock.wait();
        }
        return tRef.get();
    }

    private void runSyncOnEPHandler(final Runnable block) throws InterruptedException {
        runSyncOnEPHandler((Function<Void>) () -> {
            block.run();
            return null;
        });
    }

    @Nullable
    @Override
    public Double bufferHealth() {
        try {
            return runSyncOnEPHandler(() -> (player.getBufferedPosition() - player.getCurrentPosition()) / 1000.0);
        } catch (InterruptedException e) {
            return 0.0;
        }
    }

    @Nullable
    @Override
    public Double bufferTarget() {
        try {
            return runSyncOnEPHandler(bridge::bufferTarget);
        } catch (InterruptedException e) {
            return 0.0;
        }
    }

    @Override
    public void setBufferTarget(double target) {
        try {
            runSyncOnEPHandler(() -> bridge.setBufferTarget(target));
        } catch (InterruptedException ignored) {}
    }

    @Override
    public void setEstimatedBandwidth(@Nullable Long bps) {
        if (bps != null) {
            bandwidthMeter.setEstimatedBandwidth(bps);
        }
    }
}

File: mesh/android/ExoPlayerMesh/app/src/main/java/io/streamroot/ctl/delivery/client/mesh/exoplayermesh/PlayerInteractor.kt
package io.streamroot.ctl.delivery.client.mesh.exoplayermesh

import android.content.Context
import android.os.Handler
import android.util.Log
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.LoadControl
import com.google.android.exoplayer2.PlaybackException
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.upstream.BandwidthMeter
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DataSpec
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter
import com.google.android.exoplayer2.upstream.TransferListener
import io.streamroot.lumen.delivery.client.core.LumenPlayerInteractorBase
import io.streamroot.lumen.delivery.client.core.LumenPlayerInteractorWrapperInterface
import io.streamroot.lumen.delivery.client.core.LumenVideoPlaybackState
import java.util.concurrent.TimeUnit

typealias Driver = LumenPlayerInteractorWrapperInterface.Driver

private const val TAG = "ExoPlayerInteractor"
private const val EXO_BW_TAG = "ExoPlayerBandwidthMeter"

open class DefaultBandwidthMeterSR(context: Context) : BandwidthMeter, TransferListener {
    private val wrappedEstimator = DefaultBandwidthMeter.Builder(context).build()

    override fun getTransferListener() = this

    override fun getBitrateEstimate() = wrappedEstimator.bitrateEstimate.also {
        Log.d(EXO_BW_TAG, "[Lumen][android][DefaultBandwidthMeterSR] => Exo requested the BW, returned : $it")
    }

    override fun addEventListener(
        eventHandler: Handler,
        eventListener: BandwidthMeter.EventListener
    ) = wrappedEstimator.addEventListener(eventHandler, eventListener)

    override fun removeEventListener(eventListener: BandwidthMeter.EventListener) =
        wrappedEstimator.removeEventListener(eventListener)

    override fun onTransferInitializing(
        source: DataSource,
        dataSpec: DataSpec,
        isNetwork: Boolean
    ) = wrappedEstimator.onTransferInitializing(source, dataSpec, isNetwork)

    override fun onTransferStart(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean) =
        wrappedEstimator.onTransferStart(source, dataSpec, isNetwork)

    override fun onBytesTransferred(
        source: DataSource,
        dataSpec: DataSpec,
        isNetwork: Boolean,
        bytesTransferred: Int
    ) = wrappedEstimator.onBytesTransferred(source, dataSpec, isNetwork, bytesTransferred)

    override fun onTransferEnd(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean) =
        wrappedEstimator.onTransferEnd(source, dataSpec, isNetwork).also {
            val est = wrappedEstimator.bitrateEstimate
            Log.d(EXO_BW_TAG, "[Lumen][android][DefaultBandwidthMeterSR] => New estimated BW : $est")
        }
}

class ExoPlayerBandwidthMeter(private val context: Context, playerBuilder: ExoPlayer.Builder) : BandwidthMeter, TransferListener {
    data class HandlerAndListener(val handler: Handler, val eventListener: BandwidthMeter.EventListener)

    private val listeners = arrayListOf<HandlerAndListener>()

    private lateinit var defaultBandwidthMeterSR: DefaultBandwidthMeterSR
    private lateinit var bwDriver: LumenPlayerInteractorWrapperInterface.Driver

    private var meshEstimatedBandwidth: Long

    init {
        setDriver(Driver.PLAYER)
        meshEstimatedBandwidth = defaultBandwidthMeterSR.bitrateEstimate
        playerBuilder.setBandwidthMeter(this)
    }

    @Synchronized
    fun setDriver(driver: Driver) {
        if (this::bwDriver.isInitialized && bwDriver == driver) return
        bwDriver = driver
        Log.d(EXO_BW_TAG, "[Lumen][android][BandwidthMeter] => set BW driver : $driver")
        if (driver == Driver.PLAYER) {
            // Transfer listeners to a new BW meter
            if (this::defaultBandwidthMeterSR.isInitialized) {
                listeners.forEach { defaultBandwidthMeterSR.removeEventListener(it.eventListener) }
            }
            defaultBandwidthMeterSR = DefaultBandwidthMeterSR(context)
            listeners.forEach { defaultBandwidthMeterSR.addEventListener(it.handler, it.eventListener) }
        }
    }

    @Synchronized
    override fun getBitrateEstimate(): Long {
        val estimate = when (bwDriver) {
            Driver.PLAYER -> defaultBandwidthMeterSR.bitrateEstimate
            Driver.MESH -> meshEstimatedBandwidth
        }
        Log.d(EXO_BW_TAG, "[Lumen][android][BandwidthMeter] => Exo requested the BW, driver : ${bwDriver.name}, returned : $estimate")
        return estimate
    }

    @Synchronized
    fun setMeshEstimatedBandwidth(bps: Long) {
        meshEstimatedBandwidth = bps
        Log.d(EXO_BW_TAG, "[Lumen][android][BandwidthMeter] => New estimated BW : $bps")
    }

    override fun getTransferListener() = this

    @Synchronized
    override fun addEventListener(eventHandler: Handler, eventListener: BandwidthMeter.EventListener) {
        listeners.add(HandlerAndListener(eventHandler, eventListener))
        defaultBandwidthMeterSR.addEventListener(eventHandler, eventListener)
    }

    @Synchronized
    override fun removeEventListener(eventListener: BandwidthMeter.EventListener) {
        listeners.removeAll { it.eventListener == eventListener }
        defaultBandwidthMeterSR.removeEventListener(eventListener)
    }

    @Synchronized
    override fun onTransferInitializing(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean) {
        defaultBandwidthMeterSR.onTransferInitializing(source, dataSpec, isNetwork)
    }

    @Synchronized
    override fun onTransferStart(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean) {
        defaultBandwidthMeterSR.onTransferStart(source, dataSpec, isNetwork)
    }

    @Synchronized
    override fun onBytesTransferred(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean, bytesTransferred: Int) {
        defaultBandwidthMeterSR.onBytesTransferred(source, dataSpec, isNetwork, bytesTransferred)
    }

    @Synchronized
    override fun onTransferEnd(source: DataSource, dataSpec: DataSpec, isNetwork: Boolean) {
        defaultBandwidthMeterSR.onTransferEnd(source, dataSpec, isNetwork)
    }
}

class PlayerInteractor(
    private val player: ExoPlayer,
    loadControl: LoadControl,
    private val bandwidthMeter: ExoPlayerBandwidthMeter
) : LumenPlayerInteractorBase(),
    Player.Listener {

    init {
        player.addListener(this)
        updateBWMeter(bandwidthMeter.bitrateEstimate)
    }

    override fun onSeekProcessed() {
        super.playerStateChange(LumenVideoPlaybackState.SEEKING)
    }

    override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {
        super.playerTrackSwitch()
    }

    override fun onPlayerError(error: PlaybackException) {
        super.playerError()
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        when (playbackState) {
            Player.STATE_IDLE -> {
                super.playerStateChange(LumenVideoPlaybackState.IDLE)
            }
            Player.STATE_BUFFERING -> {
                super.playerStateChange(LumenVideoPlaybackState.REBUFFERING)
            }
            Player.STATE_READY -> {
                super.playerStateChange(if (playWhenReady) LumenVideoPlaybackState.PLAYING else LumenVideoPlaybackState.PAUSED)
            }
            Player.STATE_ENDED -> {
                super.playerStateChange(LumenVideoPlaybackState.ENDED)
            }
        }
    }

    private val handler = Handler(player.applicationLooper)

    private fun <T> runSyncOnEPHandler(block: () -> T): T? {
        var t: T? = null
        val lock = Object()
        synchronized(lock) {
            handler.post {
                synchronized(lock) {
                    t = block()
                    lock.notify()
                }
            }
            lock.wait()
        }
        return t
    }
    // Conversion from milliseconds to seconds does not use TimeUnit because we want
    // to keep a floating precision
    private val bridge = BufferTargetBridgeFactory.createInteractor(loadControl)

    override fun bufferHealth() = runSyncOnEPHandler { (player.bufferedPosition - player.currentPosition) / 1000.0 }!!

    override fun bufferTarget() = runSyncOnEPHandler { bridge.bufferTarget() }!!

    override fun setBufferTarget(target: Double) = runSyncOnEPHandler { bridge.setBufferTarget(target) }!!

    // Bandwidth

    private fun updateBWMeter(bps: Long) {
        bandwidthMeter.setMeshEstimatedBandwidth(bps)
    }
    override fun setEstimatedBandwidth(bps: Long?) {
        Log.d(EXO_BW_TAG, "[Lumen][BandwidthMeter] setEstimatedBandwidth called with value : $bps")
        bps?.let {
            updateBWMeter(bps)
        }
    }
    override fun setBandwidthDriver(driver: Driver) {
        bandwidthMeter.setDriver(driver)
    }
}

internal interface BufferTargetBridge {
    fun bufferTarget(): Double = 0.0
    fun setBufferTarget(bufferTarget: Double) {}
}

private abstract class LoadControlBufferTargetBridge(protected val loadControl: LoadControl) :
    BufferTargetBridge {

    protected fun LoadControl.getAccessibleFieldElseThrow(fieldName: String) = runCatching {
        val minBufferField = this::class.java.getDeclaredField(fieldName)
        minBufferField.isAccessible = true
        minBufferField
    }.getOrNull() ?: throw IllegalArgumentException("Impossible to retrieve field `$fieldName` value from LoadControl of type `${this::class.java.simpleName}`")

    protected fun LoadControl.getLongFromFieldElseThrow(fieldName: String) = runCatching {
        getAccessibleFieldElseThrow(fieldName).getLong(this)
    }.getOrNull() ?: throw IllegalArgumentException("Impossible to retrieve long `$fieldName` value from LoadControl of type `${this::class.java.simpleName}`")

    companion object {
        private const val MAX_BUFFER_FIELD_NAME = "maxBufferUs"
    }

    protected val maxBufferField = loadControl.getAccessibleFieldElseThrow(MAX_BUFFER_FIELD_NAME)
    protected abstract val minBufferUs: Long

    override fun bufferTarget(): Double {
        return runCatching {
            maxBufferField.getLong(loadControl).let { TimeUnit.MICROSECONDS.toSeconds(it) }.toDouble()
        }.getOrNull() ?: super.bufferTarget()
    }

    override fun setBufferTarget(bufferTarget: Double) {
        val maxBufferUs = TimeUnit.SECONDS.toMicros(bufferTarget.toLong())
        runCatching {
            maxBufferField.setLong(
                loadControl,
                maxOf(minBufferUs, maxBufferUs)
            )
        }
    }
}

// 2.10- & 2.12+
private class LoadControlBufferTargetBridgeV1(loadControl: LoadControl) :
    LoadControlBufferTargetBridge(loadControl) {
    companion object {
        private const val MIN_BUFFER_FIELD_NAME = "minBufferUs"
    }

    override val minBufferUs = loadControl.getLongFromFieldElseThrow(MIN_BUFFER_FIELD_NAME)
}

// 2.11
private class LoadControlBufferTargetBridgeV2(loadControl: LoadControl, audioOnly: Boolean) :
    LoadControlBufferTargetBridge(loadControl) {
    companion object {
        private const val MIN_BUFFER_AUDIO_FIELD_NAME = "minBufferAudioUs"
        private const val MIN_BUFFER_VIDEO_FIELD_NAME = "minBufferVideoUs"
    }

    override val minBufferUs = loadControl.getLongFromFieldElseThrow(
        if (audioOnly) MIN_BUFFER_AUDIO_FIELD_NAME else MIN_BUFFER_VIDEO_FIELD_NAME
    )
}

private object BufferTargetBridgeFactory {
    fun createInteractor(loadControl: LoadControl, audioOnly: Boolean = false): BufferTargetBridge {
        return runCatching { LoadControlBufferTargetBridgeV1(loadControl) }.getOrNull()?.also { Log.v("Misc", "[Lumen] Using interactor V1") }
            ?: runCatching { LoadControlBufferTargetBridgeV2(loadControl, audioOnly) }.getOrNull()?.also { Log.v(TAG, "[Lumen] Using interactor V2") }
            ?: throw java.lang.Exception("[Lumen] Unsupported ExoPlayer version")
    }
}